### PR TITLE
feat: implement userService (SSR, stats & profile) #246

### PR DESCRIPTION
**Pull request by PCilverio**
Implemented `userService` for profile and stats management.

Key updates:
- SSR & Client-side support for data fetching.
- Heatmap normalization logic (365-day consistency).
- Input validation for Avatar (2MB), Bio, and DisplayName.
- Account deletion safety with password confirmation.
- Full TypeScript integration and English documentation.

### CODE REVIEW

File: src/services/user/userService.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  const heatmapByDate = new Map<string, number>();
+  stats.submissionHeatmap.forEach((day: { date: string; count: number }) => {
+    heatmapByDate.set(day.date, day.count);
+  });
+
+  const today = new Date();
+  const normalizedHeatmap = Array.from({ length: 365 }, (_, index) => {
+    const date = new Date(today);
+    date.setDate(today.getDate() - (364 - index));
+    const dateKey = date.toISOString().slice(0, 10);
```
Comment (jmujica007) [inline src/services/user/userService.ts]: normalizeHeatmap() may generate off-by-one dates because toISOString() converts dates to UTC. Consider normalizing dates explicitly to avoid timezone-related issues.

File: src/services/user/userService.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    return {
+      defaultLanguage: response.data.defaultLanguage ?? DEFAULT_PREFERENCES.defaultLanguage,
+      theme: response.data.theme ?? DEFAULT_PREFERENCES.theme,
+      editorFontSize: response.data.editorFontSize ?? DEFAULT_PREFERENCES.editorFontSize,
+      editorTabSize: response.data.editorTabSize ?? DEFAULT_PREFERENCES.editorTabSize,
+    } as UserPreferences;
+  },
+
+  updateProfile: async (payload: UpdateProfilePayload): Promise<UserProfile> => {
```
Comment (jmujica007) [inline src/services/user/userService.ts:94]: Validation currently skips empty strings because of the truthy check. Consider validating explicitly against undefined instead.

File: src/services/user/userService.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    if (payload.displayName && payload.displayName.length > MAX_DISPLAY_NAME_LENGTH) {
+      throw createValidationError(`Display name exceeds ${MAX_DISPLAY_NAME_LENGTH} chars.`);
+    }
+    if (payload.bio && payload.bio.length > MAX_BIO_LENGTH) {
```
Comment (jmujica007) [inline src/services/user/userService.ts:105]: Same as displayName, this validation may skip empty string values due to the truthy condition.

File: src/services/user/userService.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      throw createValidationError(`Bio exceeds ${MAX_BIO_LENGTH} chars.`);
+    }
+
+    const response = await apiClient.put<UserProfile>('/users/me', payload);
+    return response.data;
+  },
+
+  updatePreferences: async (payload: Partial<UserPreferences>): Promise<UserPreferences> => {
```
Comment (jmujica007) [inline src/services/user/userService.ts:113]: updatePreferences() currently trusts all payload values. Consider validating theme/font settings before sending the request.

File: src/services/user/userService.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    return { avatarUrl: response.data.avatarUrl };
+  },
+
+  deleteAccount: async (payload: DeleteAccountPayload): Promise<void> => {
+    if (!payload || !payload.confirmationPassword) {
+      throw createValidationError('Confirmation password is required for account deletion.');
+    }
+
+    await apiClient.delete('/users/me', payload as unknown as RequestConfig);
```
Comment (jmujica007) [inline src/services/user/userService.ts]: Casting payload as unknown as RequestConfig bypasses type safety. It may be cleaner to extend apiClient.delete() to support request bodies explicitly.

File: src/services/user/userService.types.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  maxStreak: number;
+  submissionHeatmap: Array<{
+    date: string; // YYYY-MM-DD
+    count: number;
+  }>;
+  recentSubmissions: Submission[];
+}
+
+export interface UserPreferences {
+  defaultLanguage: string; // 'javascript' | 'python' | 'cpp' | 'java' | 'cpp'
```
Comment (jmujica007) [inline src/services/user/userService.types.ts]: defaultLanguage could use a string union type instead of plain string to improve type safety and autocomplete support.

File: src/services/user/userService.types.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
+export interface Submission {
+  id: string;
+  problemId: string;
+  problemTitle: string;
+  status: string;
+  submittedAt: string;
```
Comment (jmujica007) [inline src/services/user/userService.types.ts:14]: Submission.status might benefit from a stricter union type if the backend only supports a fixed set of statuses.

File: src/services/user/userService.test.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      });
+    });
+
+    it('throws when bio exceeds 256 characters', async () => {
+      await expect(userService.updateProfile({ bio: 'a'.repeat(257) })).rejects.toMatchObject({
+        code: 'VALIDATION_ERROR',
+      });
+    });
+  });
+});
```
Comment (jmujica007) [inline src/services/user/userService.test.ts:193]: Consider adding unit tests for invalid avatar MIME types and oversized files to fully cover the acceptance criteria.
Missing unit test for deleteAccount() without confirmationPassword, which is explicitly required by the acceptance criteria.

File: src/services/user/userService.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  editorTabSize: 2,
+};
+
+const createValidationError = (message: string): ApiError => ({
+  status: 400,
+  code: 'VALIDATION_ERROR',
+  message,
+});
+
+const createCookieConfig = (cookies?: string): RequestConfig | undefined => {
```
Comment (jmujica007) [inline src/services/user/userService.ts]: createCookieConfig() could be extracted into a shared SSR utility if similar cookie forwarding logic exists in other services.

File: src/services/user/userService.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  if (!cookies) return undefined;
+  return { headers: { Cookie: cookies } };
+};
+
+const normalizeHeatmap = (stats: UserStats): UserStats => {
+  const heatmapByDate = new Map<string, number>();
```
Comment (jmujica007) [inline src/services/user/userService.ts:49]: Heatmap normalization logic is well implemented. Consider moving it into a dedicated utility if reused across analytics/stat features later.

File: src/services/user/userService.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+  stats.submissionHeatmap.forEach((day: { date: string; count: number }) => {
+    heatmapByDate.set(day.date, day.count);
+  });
+
+  const today = new Date();
+  const normalizedHeatmap = Array.from({ length: 365 }, (_, index) => {
```
Comment (jmujica007) [inline src/services/user/userService.ts]: Consider extracting the 365 value into a named constant to improve readability and maintainability.

File: src/services/user/userService.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      '/users/me/preferences',
+      createCookieConfig(cookies)
+    );
+
+    return {
+      defaultLanguage: response.data.defaultLanguage ?? DEFAULT_PREFERENCES.defaultLanguage,
+      theme: response.data.theme ?? DEFAULT_PREFERENCES.theme,
+      editorFontSize: response.data.editorFontSize ?? DEFAULT_PREFERENCES.editorFontSize,
+      editorTabSize: response.data.editorTabSize ?? DEFAULT_PREFERENCES.editorTabSize,
+    } as UserPreferences;
```
Comment (jmujica007) [inline src/services/user/userService.ts]: Preference fallback handling looks clean. A small helper function could reduce repetition if more preference fields are added later.

File: src/services/user/userService.ts [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    const response = await apiClient.put<UserProfile>('/users/me', payload);
+    return response.data;
+  },
+
+  updatePreferences: async (payload: Partial<UserPreferences>): Promise<UserPreferences> => {
+    const response = await apiClient.put<UserPreferences>('/users/me/preferences', payload);
+    return response.data;
+  },
+
+  uploadAvatar: async (file: File): Promise<{ avatarUrl: string }> => {
```
Comment (jmujica007) [inline src/services/user/userService.ts:148]: uploadAvatar() has solid validation coverage. Consider centralizing file validation rules if other uploads are introduced in the future.

File: src/services/user/userService.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    if (!ALLOWED_AVATAR_TYPES.includes(file.type)) {
+      throw createValidationError('Invalid image format (JPEG, PNG, WEBP only).');
+    }
+    if (file.size > MAX_AVATAR_SIZE_BYTES) {
+      throw createValidationError('Image too large (Max 2MB).');
+    }
+
+    const formData = new FormData();
+    formData.append('avatar', file);
```
Comment (jmujica007) [inline src/services/user/userService.ts]: Nice use of FormData for multipart uploads. Consider defining the field name (avatar) as a constant to avoid hardcoded request keys.

File: src/services/user/userService.types.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  displayName?: string;
+  bio?: string;
+  avatarUrl?: string;
+  joinedAt: string; // ISO 8601
+  role: 'user' | 'admin';
+  badges: Badge[];
+  stats: UserStatsSummary; // lightweight summary for nav/profile card
+}
+
+export interface UserStats {
```
Comment (jmujica007) [inline src/services/user/userService.types.ts:39]: UserStats is well structured. Consider extracting the heatmap entry shape into its own reusable interface for readability.

File: src/services/user/userService.types.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  totalSolved: number;
+  easySolved: number;
+  mediumSolved: number;
+  hardSolved: number;
+  acceptanceRate: number; // percentage 0-100
+  currentStreak: number; //days
+  maxStreak: number;
+  submissionHeatmap: Array<{
```
Comment (jmujica007) [inline src/services/user/userService.types.ts]: Inline object types work fine here, but a dedicated HeatmapEntry interface could improve reusability and readability.

File: src/services/user/userService.test.ts [hunk 17]
```diff
@@ -0,0 +1,99 @@
+import { describe, expect, it, vi, beforeEach } from 'vitest';
+
+import { userService } from './userService';
+
+const mockApiResponse = <T>(data: T) => ({ data });
+
+describe('userService', () => {
```
Comment (jmujica007) [inline src/services/user/userService.test.ts:7]: Test setup looks clean. You could also consider extracting common mock responses into reusable fixtures as the test suite grows.

File: src/services/user/userService.test.ts [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        status: 200,
+        json: async () => mockApiResponse(profile),
+      } as Response);
+
+      const result = await userService.getProfile('2');
+      expect(result.username).toBe('sofi');
+    });
+  });
+
+  describe('getStats', () => {
```
Comment (jmujica007) [inline src/services/user/userService.test.ts:58]: Good separation of test suites by feature. This structure will scale well as more service methods are added.

File: src/services/user/userService.ts (hunk 9 above)
Comment (PCilverio) [reply src/services/user/userService.ts]: Good idea! You're right that it belongs to the global API utilities. However, to avoid modifying files outside the scope of this task, I'll leave it local for now. I exported it so it's easy to move in a future global refactor.

File: src/services/user/userService.ts (hunk 11 above)
Comment (PCilverio) [reply src/services/user/userService.ts]: Absolutely! I've extracted the value 365 into a constant named DAYS_IN_YEAR at the beginning of the file to eliminate the magic number and make the code more maintainable.

File: src/services/user/userService.ts (hunk 1 above)
Comment (PCilverio) [reply src/services/user/userService.ts]: You're absolutely right. To avoid the time zone offset caused by toISOString(), I implemented a helper function, formatDateLocal(), which extracts the year, month, and day from the instance directly in local time. Thanks for pointing out that critical detail!

File: src/services/user/userService.ts (hunk 12 above)
Comment (PCilverio) [reply src/services/user/userService.ts]: Excellent observation! I applied a refactor using the propagation operator (...response.data) on the DEFAULT_PREFERENCES. This makes the code completely DRY, scalable for future fields, and eliminates the need to manually map property by property.

File: src/services/user/userService.test.ts (hunk 8 above)
Comment (PCilverio) [reply src/services/user/userService.test.ts:193]: Fixed! I've added alternate test cases for invalid MIME formats and sizes exceeding the 2MB limit in uploadAvatar. I've also included the error test for when deleteAccount runs without confirmationPassword. Thanks for ensuring 100% coverage!

File: src/services/user/userService.test.ts (hunk 18 above)
Comment (PCilverio) [reply src/services/user/userService.test.ts:58]: Thank you so much! Keeping the blocks clearly defined for each method helps us quickly map acceptance criteria and scale tests smoothly.

File: src/services/user/userService.test.ts (hunk 17 above)
Comment (PCilverio) [reply src/services/user/userService.test.ts:7]: Good point! You're right that reusing mock profiles will prevent code duplication as we add more tests. I'll keep that in mind as an improvement for the next refactor.

File: src/services/user/userService.ts [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+      date: dateKey,
+      count: heatmapByDate.get(dateKey) ?? 0,
+    };
+  });
+
+  return { ...stats, submissionHeatmap: normalizedHeatmap };
+};
+
+// Service functions
+export const userService = {
```
Comment (Ditmar) [inline src/services/user/userService.ts:69]: The validation for the `confirmationPassword` in the `deleteAccount` method checks if it's provided, but it does not validate its length or content. This could lead to silent failures if an empty string is passed.

**Suggestion:**
```typescript
Consider adding a check to ensure that the `confirmationPassword` is not only present but also meets certain criteria (e.g., minimum length).
```

File: src/services/user/userService.ts [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+      '/users/me/preferences',
+      createCookieConfig(cookies)
+    );
+
+    return {
+      ...DEFAULT_PREFERENCES,
+      ...response.data,
+    };
+  },
+
```
Comment (Ditmar) [inline src/services/user/userService.ts:93]: The `updatePreferences` method does not validate the `defaultLanguage` field, which could lead to invalid values being accepted.

**Suggestion:**
```typescript
Add validation to ensure that `defaultLanguage` is one of the supported languages.
```

File: src/services/user/userService.ts [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+  updateProfile: async (payload: UpdateProfilePayload): Promise<UserProfile> => {
+    if (payload.displayName !== undefined && payload.displayName.length > MAX_DISPLAY_NAME_LENGTH) {
```
Comment (Ditmar) [inline src/services/user/userService.ts:99]: The `updatePreferences` method allows for partial updates, but it does not handle cases where none of the fields are provided, which could lead to unintended behavior.

**Suggestion:**
```typescript
Consider adding a check to ensure at least one field is provided in the payload.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
The implementation of the userService module is well-structured and adheres to the requirements outlined in the ticket. The tests cover various scenarios, ensuring that the service functions correctly. However, there are some missing validations and edge cases that need to be addressed for robustness. Overall, the code is maintainable and follows good practices, but attention to detail in validation is necessary.
---

**Conclusion:** The implementation is generally solid, but critical validation checks are missing, which could lead to security and correctness issues. These must be addressed before merging.

File: src/services/user/userService.ts [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+    date.setDate(today.getDate() - (DAYS_IN_YEAR - 1 - index));
+    const dateKey = formatDateLocal(date);
+    return {
+      date: dateKey,
+      count: heatmapByDate.get(dateKey) ?? 0,
+    };
+  });
+
+  return { ...stats, submissionHeatmap: normalizedHeatmap };
+};
```
Comment (Ditmar) [inline src/services/user/userService.ts:66]: The validation for the displayName length does not provide a specific error message when the length exceeds the limit, which could lead to confusion.

**Suggestion:**
```typescript
Consider adding a specific error message for when the displayName exceeds the maximum length.
```

File: src/services/user/userService.ts [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+
+// Service functions
+export const userService = {
+  getProfile: async (userId?: string, cookies?: string): Promise<UserProfile> => {
+    const path = userId ? `/users/${userId}` : '/users/me';
+    const response = await apiClient.get<UserProfile>(path, createCookieConfig(cookies));
+    return response.data;
+  },
+
```
Comment (Ditmar) [inline src/services/user/userService.ts:75]: The validation for the bio length does not provide a specific error message when the length exceeds the limit, which could lead to confusion.

**Suggestion:**
```typescript
Consider adding a specific error message for when the bio exceeds the maximum length.
```

File: src/services/user/userService.ts [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+  updatePreferences: async (payload: Partial<UserPreferences>): Promise<UserPreferences> => {
+    if (!payload || Object.keys(payload).length === 0) {
+      throw createValidationError('Payload cannot be empty.');
+    }
+
+    if (
+      payload.defaultLanguage !== undefined &&
+      !ALLOWED_LANGUAGES.includes(payload.defaultLanguage)
+    ) {
```
Comment (Ditmar) [inline src/services/user/userService.ts:121]: The validation for the editorFontSize does not provide a specific error message when the value is out of range, which could lead to confusion.

**Suggestion:**
```typescript
Consider adding a specific error message for when the editorFontSize is below 12 or above 24.
```

File: src/services/user/userService.ts [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+
+    if (
+      payload.editorFontSize !== undefined &&
+      (payload.editorFontSize < 12 || payload.editorFontSize > 24)
+    ) {
+      throw createValidationError(`Font size must be between 12 and 24.`);
+    }
+
+    if (payload.theme !== undefined && !['light', 'dark', 'system'].includes(payload.theme)) {
+      throw createValidationError('Invalid theme selection.');
```
Comment (Ditmar) [inline src/services/user/userService.ts:133]: The validation for the editorTabSize does not provide a specific error message when the value is invalid, which could lead to confusion.

**Suggestion:**
```typescript
Consider adding a specific error message for when the editorTabSize is not 2 or 4.
```

File: src/services/user/userService.ts [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+    if (
+      payload.editorTabSize !== undefined &&
+      payload.editorTabSize !== 2 &&
+      payload.editorTabSize !== 4
+    ) {
+      throw createValidationError('Tab size must be either 2 or 4.');
+    }
+
+    const response = await apiClient.put<UserPreferences>('/users/me/preferences', payload);
+    return response.data;
```
Comment (Ditmar) [inline src/services/user/userService.ts:145]: The validation for the avatar file format does not specify which formats are allowed, which could lead to confusion.

**Suggestion:**
```typescript
Consider including the allowed formats in the error message for clarity.
```

File: src/services/user/userService.ts [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+  },
+
+  uploadAvatar: async (file: File): Promise<{ avatarUrl: string }> => {
+    if (!ALLOWED_AVATAR_TYPES.includes(file.type)) {
+      throw createValidationError('Invalid image format (JPEG, PNG, WEBP only).');
+    }
+    if (file.size > MAX_AVATAR_SIZE_BYTES) {
+      throw createValidationError('Image too large (Max 2MB).');
```
Comment (Ditmar) [inline src/services/user/userService.ts]: The validation for the avatar file size does not specify the maximum size allowed in the error message, which could lead to confusion.

**Suggestion:**
```typescript
Consider including the maximum size in the error message for clarity.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
The code introduces a comprehensive test suite for the user service, which is a positive addition for maintainability and correctness. However, there are some areas that could be improved for better robustness and clarity, particularly around error handling and validation logic.
---

**Conclusion:** While the test suite is a strong addition, the error handling and validation messages need to be more specific to improve user experience and maintainability.

**Review by jmujica007 — APPROVED** [review]

**Review by Ditmar — APPROVED** [review]
